### PR TITLE
Multicluster config kubernetes updates

### PR DESCRIPTION
**Pull request by timmyreilly**
Updating kubernetes client to work with multicluster config

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [X] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change?

### CODE REVIEW

File: pkg/kubernetes/types.go [hunk 1]
```diff
@@ -103,4 +108,10 @@ type Controller interface {
 
 	// IsMetricsEnabled returns true if the pod in the mesh is correctly annotated for prometheus scrapping
 	IsMetricsEnabled(*corev1.Pod) bool
+
+	// GetMultiClusterSvc returns an v1alpha1 MulticClusterService if it exists, or nil
+	GetMultiClusterSvc(name, namespace string) (*v1alpha1.MultiClusterService, error)
+
+	// ListMultiClusterServices returns a list of v1alpha1 MulticClusterServices if any exist, or nil
+	ListMultiClusterServices(name, namespace string) ([]*v1alpha1.MultiClusterService, error)
```
Comment (timmyreilly) [inline pkg/kubernetes/types.go:116]: remove name parameter, all good with namespace?

Comment (timmyreilly) [reply pkg/kubernetes/types.go:116]: Update to "all" by removing all parameters

File: pkg/kubernetes/client.go [hunk 2]
```diff
@@ -311,3 +330,28 @@ func (c Client) IsMetricsEnabled(pod *corev1.Pod) bool {
 	isScrapingEnabled, _ = strconv.ParseBool(prometheusScrapeAnnotation)
 	return isScrapingEnabled
 }
+
+// GetMultiClusterSvc returns an v1alpha1 MulticClusterService if it exists, or nil
+func (c Client) GetMultiClusterSvc(name, namespace string) (*v1alpha1.MultiClusterService, error) {
```
Comment (snehachhabria) [inline pkg/kubernetes/client.go:335]:
```suggestion
func (c Client) GetMultiClusterService(name, namespace string) (*v1alpha1.MultiClusterService, error) {
```

File: pkg/kubernetes/client.go [hunk 3]
```diff
@@ -311,3 +330,28 @@ func (c Client) IsMetricsEnabled(pod *corev1.Pod) bool {
 	isScrapingEnabled, _ = strconv.ParseBool(prometheusScrapeAnnotation)
 	return isScrapingEnabled
 }
+
+// GetMultiClusterSvc returns an v1alpha1 MulticClusterService if it exists, or nil
```
Comment (snehachhabria) [inline pkg/kubernetes/client.go:334]:
```suggestion
// GetMultiClusterService returns an v1alpha1 MultiClusterService if it exists, or nil
```

File: pkg/kubernetes/client.go [hunk 4]
```diff
@@ -311,3 +330,28 @@ func (c Client) IsMetricsEnabled(pod *corev1.Pod) bool {
...
+func (c Client) GetMultiClusterSvc(name, namespace string) (*v1alpha1.MultiClusterService, error) {
+	svcIf, exists, err := c.informers[MultiClusterServices].GetStore().GetByKey(name)
+	if exists && err == nil {
+		mcsvc := svcIf.(*v1alpha1.MultiClusterService)
+		return mcsvc, nil
+	}
+	return nil, nil
+}
+
+// ListMultiClusterServices returns a list of v1alpha1 MulticClusterServices if any exist, or nil
```
Comment (snehachhabria) [inline pkg/kubernetes/client.go:344]:
```suggestion
// ListMultiClusterServices returns a list of v1alpha1 MultiClusterServices if any exist, or nil
```

File: pkg/announcements/types.go [hunk 5]
```diff
@@ -156,6 +156,17 @@ const (
...
+	// ---
+
+	// MultiClusterServiceAdded is the type of announcement emitted when we observe an addition of a multiclusterservice.config.openservicemesh.io
+	MultiClusterServiceAdded AnnouncementType = "multiclusterservice-added"
+
+	// MultiClusterServiceDeleted is the type of announcement emitted when we observe an deletion of a multiclusterservice.config.openservicemesh.io
+	MultiClusterServiceDeleted AnnouncementType = "multiclusterservice-deleted"
+
+	// MultiClusterServiceUpdated is the type of announcement emitted when we observe an update of a multiclusterservice.config.openservicemesh.io
+	MultiClusterServiceUpdated AnnouncementType = "multiclusterservice-updated"
```
Comment (snehachhabria) [inline pkg/announcements/types.go:169]: You will need to add this to this file as well : https://github.com/openservicemesh/osm/blob/113bfc328b857272dc149066d162fc950f0b584f/pkg/catalog/dispatcher.go#L43

File: pkg/kubernetes/client.go [hunk 6]
```diff
@@ -98,6 +102,21 @@ func (c *Client) initServicesMonitor() {
 	c.informers[Services].AddEventHandler(GetKubernetesEventHandlers((string)(Services), providerName, c.shouldObserve, svcEventTypes))
 }
 
+func (c *Client) initMultiClusterServicesMonitor() {
```
Comment (snehachhabria) [inline pkg/kubernetes/client.go:105]: Adding the changes here doesn't seem right, if I look at both the other custom crd's that we have in osm the changes for their informers are in these files : 

1. egress crd : https://github.com/openservicemesh/osm/blob/113bfc328b857272dc149066d162fc950f0b584f/pkg/policy/client.go#L41
2. meshConfig crd : https://github.com/openservicemesh/osm/blob/113bfc328b857272dc149066d162fc950f0b584f/pkg/configurator/client.go#L29

Please could we follow these examples and make the change for the mulitcluster crd accordingly

**Review by snehachhabria — CHANGES_REQUESTED** [review]

File: pkg/kubernetes/client.go (hunk 6 above)
Comment (steeling) [reply pkg/kubernetes/client.go:105]: I'd argue that those ways are more incorrect for the following reasons:

1. It places more direct dependencies on the MeshCatalog to Kubernetes (these should be abstracted through providers).
2. It's kind of messy to have a separate client for each CRD.. code will grow unwieldy
3. Knowledge of CRD's will become more intertwined, meaning more types will need to reconcile MultiClusterService with Service, or other CRD's. Forcing each type to need a client for each of these CRD's (when it's all reaching out to the Kuberentes master) is a burden on the developer. For a concrete example, we would then need to modify the kube provider to also have a MultiClusterServiceController.
4. Clients are clients that talk to a "thing". In this case that thing is the K8s master, why have 9 clients for 9 of the resources backed by that same "thing".

If you'd like, we can pick up the work to remove those 2 clients and migrate the code to the Kubernetes Controller.

File: pkg/kubernetes/client.go [hunk 7]
```diff
@@ -32,11 +35,12 @@ func NewKubernetesController(kubeClient kubernetes.Interface, meshName string, s
```
Comment (steeling) [inline pkg/kubernetes/client.go:35]: you'll need the kubernetes controller to accept and populate the ConfigInterface

File: pkg/kubernetes/client.go (hunk 6 above)
Comment (steeling) [reply pkg/kubernetes/client.go:105]: https://github.com/openservicemesh/osm/pull/3635 brings in those changes for the other client types

**Pull request closed without merging**